### PR TITLE
add highs solver

### PR DESCRIPTION
**Pull request by pz-max**
Closes # (if applicable).

## Changes proposed in this Pull Request
Add open source [HiGHS](https://github.com/ERGO-Code/HiGHS) solver:
solver_performance_new

Future work can:
- While the above results look incredible, generally all open source solvers struggle when 1 year is simulated in hourly resolution. I.e. Gurobi solves 5 nodes for 1 year in hourly resolution in 1min, while HiGHS and others may need 2+ hours. This performance drop occurs because open source solvers don't address properly the existence of "dense columns".
- Future work can contribute to the HiGHS solver and implement a method for the IPM that does deal with dense columns.
- Other work could implement (advanced) aggressive scaling techniques into PyPSA (to be useful for all solvers) and/or HiGHS. See some ongoing PyPSA discussion here: https://github.com/PyPSA/PyPSA/issues/309

## Checklist

- [x] Code changes are sufficiently documented; i.e. new functions contain docstrings and further explanations may be given in `doc`.
- [x] A note for the release notes `doc/release_notes.rst` of the upcoming release is included.
- [x] I consent to the release of this PR's code under the MIT license.

### CODE REVIEW

**Comment by pz-max** [comment]
Will add some other features, like using threads in parallel. Just going through tests.

File: pypsa/linopt.py [hunk 1]
```diff
@@ -601,6 +601,142 @@ def set_int_index(ser):
...
+    variables_sol : series
+    constraints_dual : series
+    objective : float
+    """
+    import logging, re, io, subprocess, sys, os
+
+    # write option file
+    if not os.path.exists(solution_fn):
+        os.mknod(solution_fn)
+    f1 = open("highs_options.txt","w")
```
Comment (FabianHofmann) [inline pypsa/linopt.py:668]: this should probably be a temporary file

**Review by FabianHofmann — DISMISSED** [review]
This is really nice! 
I think the solver options could be handled in a better way. A suggestion:

1. pop all the command line options from the `solver_options` dict with defaults. In the form `parallel = solver_options.pop(parallel, False)`. All of these are added to the command line (later)
2. Set a defaults dict for the `option_file_options` which include the options which you have in line 670-674. 
3. Update the `option_file_options` dict with the remaining `solver_options`, using `option_file_options.update(solver_options)` 
4. Convert the `option_file_options` to text form and write it to the file. 

In this way, one can easily access command line options and option-file options. What do you think?

**Review by FabianHofmann — COMMENTED** [review]
That looks good! Let me go one time through the installation instruction and run a optimization. From the code everything looks fine. 

Would you mind if I copy some snippets to the linopy package? Otherwise you could just copy/paste that function into the linopy/solvers.py (so it stays effectively your contribution) and make a PR and the rest I take care of.

**Comment by pz-max** [comment]
HI @FabianHofmann, We are testing the installation today with @fneum. Yeah, I would write a PR in Linopy as well. Probably good to have some overview on the codebase anyways :)

File: pypsa/linopt.py [hunk 2]
```diff
@@ -601,6 +601,165 @@ def set_int_index(ser):
...
+    solution and dual constraint variables. Highs must be installed for usage.
+
+    Installation
+    -------------
+    Installation manual: https://www.maths.ed.ac.uk/hall/HiGHS/
+    After the "make" run, the 'highs' executables are in highs/build/bin/run/highs
+    Once the "ctest" runs make sure you set the path. Path setting notes:
+    >>> highs lib folder must be in "LD_LIBRARY_PATH" environment variable
+    >>> To do this, insert the below path in the .bashrc (hidden in the home folder in Linux/Ubuntu)
+    >>> LD_LIBRARY_PATH=$LD_LIBRARY_PATH:/<your_path>/HiGHS/build/lib
```
Comment (fneum) [inline pypsa/linopt.py:619]: also the HiGHS executables need to be added to the path. Here are the full installation steps I needed.

```bash
sudo apt-get install cmake # if not installed
git clone git@github.com:ERGO-Code/HiGHS.git
cd HiGHS
mkdir build
cd build
cmake ..
make
ctest
```

Then in `.bashrc` add paths of executables and library:

```sh
export PATH="${PATH}:/foo/HiGHS/build/bin"
export LD_LIBRARY_PATH="${LD_LIBRARY_PATH}:/foo/HiGHS/build/lib"
```

```sh
source .bashrc
```

Now you should see when running `highs`:

```
Running HiGHS 1.1.1 [date: 2021-11-11, git hash: 95342daa]
```

**Comment by pz-max** [comment]
some to do's https://github.com/PyPSA/linopy/pull/8#issuecomment-966485362

**Comment by pz-max** [comment]
Hi guys,
- I installed the new HiGHs version
- Switched the options to `"write_solution_style": 1,` where output variables good be better read out
- Tested with PyPSA
- Tested with PyPSA-Eur

Some outputs for the PyPSA run:
Solver outputs for variable primal:
comparison_solvers1

Solver outputs for constraints dual:
constraint_dual

The difference for the PyPSA-Eur run compared to Gurobi are also minor. So seems now everything works.

**Comment by fneum** [comment]
Thanks @pz-max for driving this! I would like to also see that the variables are correctly attributed to the same variables. The stats somewhat show that all the values of the variables are there but not whether they are correctly allocated.

**Comment by pz-max** [comment]
Luckily I was taking this snapshot:
image

The new `"write_solution_style": 1,` lead to an output with variable naming such as x300 = ... as the other solvers do. I had to use also `.pipe(set_int_index)` for the variables to associate them right

**Comment by pz-max** [comment]
@fneum, here a more complete version:
[solution_variables.csv](https://github.com/PyPSA/PyPSA/files/7547208/solution_variables.csv)
Note, the index had before cleaning an "x" in front of the number i.e. "x300", meaning that the variables should be matched now.

**Comment by fneum** [comment]
What are the column labels for the .csv? Are the objective values the same? Looking mostly good, but could you investigate whether the differences are degeneracies?

**Comment by pz-max** [comment]
Hi @fneum, Thanks for that note. Now all solutions with column index:
- [objective.csv](https://github.com/PyPSA/PyPSA/files/7561668/objective.csv)
- [constraint_dual.csv](https://github.com/PyPSA/PyPSA/files/7561666/constraint_dual.csv)
- [solution_variables.csv](https://github.com/PyPSA/PyPSA/files/7561669/solution_variables.csv)

I checked the outcomes of the results for 3 runs (degeneracy check I assume?). We can observe always the same difference/results. The differences of the unique results are given here:
image
image

Overall the differences are not large. Turning off the crossover, didn't change anything (would need to check the C code if anything was triggered). What are your thoughts? Should we find out how to reduce the little differences or is it already ok?

**Comment by fneum** [comment]
The comparison looks almost all very good! The only thing that bothers me is the 1250 difference from variable x270 I presume. Could you check what variable this is and what cost is associated with it? If there is no cost associated with this variable it's fine.

File: pypsa/linopt.py [hunk 3]
```diff
@@ -601,6 +601,175 @@ def set_int_index(ser):
...
+    you could write a dictionary replacing some of the default values or adding new
+    options:
+    ```
+    solving:
+        solver:
+            name: highs,
+            method: ipm,
+            parallel: "on",
+            <option_name>: <value>,
+    ```
```
Comment (fneum) [inline pypsa/linopt.py:673]: This explanation should not refer to PyPSA-Eur. Write how you can use the `solver_options` keyword.

```py
solver_options = {
    "method": "ipm",
    ...
}
```

File: pypsa/linopt.py [hunk 4]
```diff
@@ -601,6 +601,175 @@ def set_int_index(ser):
...
+    Note, the <option_name> and <value> must be equivalent to the name convention
+    of HiGHS. Some function exist that are not documented, check their GitHub file:
+    \HighsOptions.h
```
Comment (fneum) [inline pypsa/linopt.py:676]: Can you provide full link?

File: pypsa/linopt.py [hunk 5]
```diff
@@ -601,6 +601,175 @@ def set_int_index(ser):
...
+    constraints_dual : series
+    objective : float
+    """
+    options_fn = "highs_options.txt"
+    default_dict = {
+        "method": "ipm",
+        "primal_feasibility_tolerance": 1e-04,
+        "dual_feasibility_tolerance": 1e-05,
+        "ipm_optimality_tolerance": 1e-6,
+        # "ipm_iteration_limit": 100,
```
Comment (fneum) [inline pypsa/linopt.py:694]: Remove if commented out.

File: pypsa/linopt.py [hunk 6]
```diff
@@ -601,6 +601,175 @@ def set_int_index(ser):
...
+    # read out solution file (.sol)
+    f = open(solution_fn, "rb")
+    trimed_sol_fn = re.sub(rb'\*\*\s+', b'', f.read())
+    f.close()
+    
+    sol = pd.read_csv(io.BytesIO(trimed_sol_fn), header=[1], sep=r'\s+')
+    row_no = sol[sol["Index"] == 'Rows'].index[0]
+    sol = sol.drop(row_no+1)  # Removes header line after "Rows"
+    sol_rows = sol[(sol.index > row_no)]
+    sol_cols = sol[(sol.index < row_no)].set_index("Name").pipe(set_int_index)
```
Comment (fneum) [inline pypsa/linopt.py:764]: Wouldn't it be possible to use `set_int_index` as well for `sol_rows`. Then you wouldn't need the `reset_index(drop=True) + 1` below.

**Comment by fneum** [comment]
I think these are my final comments, after that, good to go in.

File: doc/release_notes.rst [hunk 7]
```diff
@@ -16,6 +16,7 @@ Upcoming Release
 
 PyPSA 0.18.1 (15th October 2021)
 ================================
+* Add "highs" solver.
```
Comment (fneum) [inline doc/release_notes.rst:19]: Let's add a link.

**Comment by pz-max** [comment]
Hi @fneum,
https://github.com/PyPSA/PyPSA/pull/308#issuecomment-974151807. Attached the lp file:
[lpfile_ac_dc_meshed_example.txt](https://github.com/PyPSA/PyPSA/files/7574902/lpfile_ac_dc_meshed_example.txt)

Attached the .lp file for the above solutions. I analyzed here the pypsa ac_dc example mesh. From the .lp file as well as from .solution_primal_variable file above, we can see that there is no value assigned to **x270**. I guess this means x270 = 0 (never read out variables, hopefully I interpret that correctly).

File: pypsa/linopt.py (hunk 6 above)
Comment (pz-max) [reply pypsa/linopt.py:764]: This does not work. Below is the dataframe of `sol_rows`. While sol_cols have variables with a "x" i.e. x300, `sol_rows` doesn't have that. Therefore, set_int_index does not work (could be modified).

image

Linopy will be added soon, I think it's probably not useful to modify the set_int_index function?

File: pypsa/linopt.py (hunk 4 above)
Comment (pz-max) [reply pypsa/linopt.py]: done

File: pypsa/linopt.py (hunk 5 above)
Comment (pz-max) [reply pypsa/linopt.py]: done in https://github.com/PyPSA/PyPSA/pull/308/commits/9becfe64b6a12c6d1f9640d923b431192b2073d4

File: pypsa/linopt.py (hunk 6 above)
Comment (fneum) [reply pypsa/linopt.py:762]: That sounds absolutely reasonable. Thought the duals may have a "y" prefix.

**Comment by fneum** [comment]
> Attached the .lp file for the above solutions. I analyzed here the pypsa ac_dc example mesh. From the .lp file as well as from .solution_primal_variable file above, we can see that there is no value assigned to x270. I guess this means x270 = 0 (never read out variables, hopefully I interpret that correctly).

The variable x270 does show up in both LP file and solution. For gurobi, CBC it's 50, whereas for highs it's 1200. That's curious. What variable is it? Investment variable of a generator or link?

**Comment by pz-max** [comment]
Hi @fneum, I was looking for it. How/where can I check this out?

**Comment by pz-max** [comment]
> Hi @fneum, I was looking for it. How/where can I check this out?

**Update.** When `network_lopf` is run with `keep_references=True` as described [here](https://pypsa.readthedocs.io/en/latest/api_reference.html?highlight=network_lopf#pypsa.linopf.network_lopf) then the variables/constraints can be checked. We found out that in the example above the problematic variable x=270 is related to the state_of_charge variable. 

image

While Gurobi, Cbc and Glpk are solving towards the same variable (x270=100), only HiGHs slightly deviate (x270=1250). This suggests that something is going wrong in the HiGHS solver. The reasons are unclear, therefore a **logger.warning** will be added. Nevertheless, all implemented solvers are very accurate regarding the objective solution (objective=14337113. €) with no deviation before the comma.

**Comment by fneum** [comment]
Congratulations on this very nice contribution! Merging whenever you are ready.

**Review by fneum — APPROVED** [review]

**Pull request merged**